cli/lsp/sources.rs

```diff
@@ -40,16 +40,108 @@ pub async fn cache(
   builder.add(specifier, false).await
 }
 
+fn get_remote_headers(
+  cache_filename: &Path,
+) -> Option<HashMap<String, String>> {
+  let metadata_path = http_cache::Metadata::filename(cache_filename);
+  let metadata_str = fs::read_to_string(metadata_path).ok()?;
+  let metadata: http_cache::Metadata =
+    serde_json::from_str(&metadata_str).ok()?;
+  Some(metadata.headers)
+}
+
+fn resolve_remote_specifier(
+  specifier: &ModuleSpecifier,
+  http_cache: &HttpCache,
+  redirect_limit: isize,
+) -> Option<ModuleSpecifier> {
+  let cache_filename = http_cache.get_cache_filename(specifier.as_url());
+  if redirect_limit >= 0 && cache_filename.is_file() {
+    let headers = get_remote_headers(&cache_filename)?;
+    if let Some(location) = headers.get("location") {
+      let redirect =
+        ModuleSpecifier::resolve_import(location, specifier.as_str()).ok()?;
+      resolve_remote_specifier(&redirect, http_cache, redirect_limit - 1)
+    } else {
+      Some(specifier.clone())
+    }
+  } else {
+    None
+  }
+}
+
+fn resolve_specifier(
+  specifier: &ModuleSpecifier,
+  redirects: &mut HashMap<ModuleSpecifier, ModuleSpecifier>,
+  http_cache: &HttpCache,
+) -> Option<ModuleSpecifier> {
+  let scheme = specifier.as_url().scheme();
+  if !SUPPORTED_SCHEMES.contains(&scheme) {
+    return None;
+  }
+
+  if scheme == "data" {
+    Some(specifier.clone())
+  } else if scheme == "file" {
+    let path = specifier.as_url().to_file_path().ok()?;
+    if path.is_file() {
```

Comment:
This looks susceptible to TOCTOU bugs, the on-disk entity may change between check and use.

Reply:
Can you think of an easy way to mitigate it?

The scenario it would arise is pretty rare, in that this is a file on disk that would need to be modified in a few milliseconds by something other than the current editor and the impact is that the lsp might have a slightly out of date version of a document until some other change causes the disconnect to be rectified.